CSIStorageCapacity beta API

pohly opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

We want to promote CSIStorageCapacity to beta in Kubernetes 1.21.

#### Special notes for your reviewer:

The new v1beta1 API might get extended before Kubernetes 1.21 gets released if the new CSI spec 1.4.0 gets released quickly enough. Right now, this PR does not include that change (yet).

#### Does this PR introduce a user-facing change?
```release-note
storage capacity tracking (= the CSIStorageCapacity feature) is beta, storage.k8s.io/v1alpha1/VolumeAttachment and storage.k8s.io/v1alpha1/CSIStorageCapacity objects are deprecated
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1472-storage-capacity-tracking
- [Usage]: https://kubernetes.io/docs/concepts/storage/storage-capacity/
```

pohly committed:
CSIStorageCapacity: promote API to beta

The v1alpha1 API is left in place for now to ease the migration.

pohly committed:
CSIStorageCapacity: use beta API

pohly committed:
CSIStorageCapacity: generated files for beta

pohly committed:
CSIStorageCapacity: prepare tests for enabling the feature by default

Defaults and validation are such that the field has to be set when
the feature is enabled, just as for the other boolean fields. This
was missing in some tests, which was okay as long as they ran
with the feature disabled. Once it gets enabled, validation will
flag the missing field as error.

Other tests didn't run at all.

pohly committed:
CSIStorageCapacity: update rbac test data

This is the result of
  UPDATE_BOOTSTRAP_POLICY_FIXTURE_DATA=true go test k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy
after enabling the CSIStorageCapacity feature. This enables
additional RBAC entries for reading CSIDriver and
CSIStorageCapacity.

pohly committed:
CSIStorageCapacity: mark v1alpha1 as deprecated

We still need the alpha API for testing with current
external-provisioner, but then should remove it shortly after the 1.21
release once the external-provisioner is updated to use the beta API.

That VolumeAttachment is still defined in the alpha API looks like an
oversight. To minimize changes during the beta graduation of
CSIStorageCapacity it is left in place with a deprecation notification
in 1.21.

pohly commented on test/integration/etcd/data.go:
```diff
@@ -322,6 +333,16 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
+			// TODO: should this be beta1?
+			// alpha is what makes `go test k8s.io/kubernetes/test/integration/etcd` pass.
+			ExpectedGVK: gvkP("storage.k8s.io", "v1alpha1", "CSIStorageCapacity"),
+		},
```
I could use some guidance. What is the right value for `ExpectedGVK` here? It passes `go test k8s.io/kubernetes/test/integration/etcd` when using "v1alpha1", but that seems counter-intuitive. Do I perhaps have some missing code or a bug elsewhere such that "v1beta1" is not used?

When bringing up a cluster, the v1beta1 API works as expected. A writer which stores CSIStorageCapacity objects using v1alpha1 works and then `kubect get csistoragecapacities` returns the objects as v1beta1.

liggitt replied:
Yes, this means the data is still being stored in etcd in v1alpha1.

`NewStorageFactoryConfig()` is explicitly overriding the preferred storage version for this type to be v1alpha1.

pohly replied:
Are these overrides in `NewStorageFactoryConfig()` needed? I probably added the one for `csistoragecapacities` only because there was one for `csidrivers`, something that got removed since then.

liggitt replied:
yes. v1 is the preferred etcd version for storage.k8s.io, but this type does not exist in that version yet.

pohly replied:
Looks like these overrides are necessary for all (?) APis which are not yet in v1. I'll change the existing entry for "csistoragecapacities" to "v1beta1"

liggitt commented on test/integration/etcd/data.go:
```diff
@@ -269,6 +269,17 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
 		// --
 
+		// k8s.io/kubernetes/pkg/apis/storage/v1alpha1
+		gvr("storage.k8s.io", "v1alpha1", "csistoragecapacities"): {
```
it is surprising this was not required before. it looks like the API registration was contingent on `if utilfeature.DefaultFeatureGate.Enabled(features.CSIStorageCapacity)`, which is unusual. Enabling the alpha API with `--runtime-config=storage.k8s.io/v1alpha1=true` is typically all that is required. Adding the feature gate check meant that this test didn't exercise storage path and version testing

pohly replied:
> it looks like the API registration was contingent on if utilfeature.DefaultFeatureGate.Enabled(features.CSIStorageCapacity), which is unusual.

What if the v1alpha1 API group had had two different APIs which are under different feature gates? From that perspective it made sense to me to toggle the API individually based on the feature gate.

What about the REST storage provider? There basically all API types get added depending on the feature gates, for example:
https://github.com/kubernetes/kubernetes/blob/603a2aa8a99869dc51ea1be2ce6e6ca10c23e55f/pkg/registry/storage/rest/storage_storage.go#L98-L105

Is that normal or unusual?

But I agree, this did have the undesired effect that several tests simply didn't run for the CSIStorageCapacity object before.

liggitt replied:
> What if the v1alpha1 API group had had two different APIs which are under different feature gates? From that perspective it made sense to me to toggle the API individually based on the feature gate.
> 
> What about the REST storage provider? There basically all API types get added depending on the feature gates, for example:
> ...
> Is that normal or unusual?

That's unusual. Opting into enabling an alpha API version typically gets you everything in the version. Disabling beta APIs is typically done at the version level.

pohly commented:
Rebased because of conflict in `test/integration/apiserver/print_test.go`.

pohly commented:
> pull-kubernetes-verify — Job failed.
>...
> /home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh

Looks like something changed in master. I rebased and then could regenerate the openapi file. Pushed.

pohly commented:
The PR should now be in a state were it could be merged. What's missing is a potential extension of the CSIStorageCapacity object with a `MaximumVolumeSize` field that will come with a CSI spec update to 1.4.0, which may or may happen in time for Kubernetes 1.21.

My preference would be to get this reviewed and merged as-is if we get the review done soonish. The follow-up PR then will be smaller. The downside is that we then need to get two PRs through API review.

We can also hold this one until the CSI spec update is included. Please let me know if that is what you prefer.

/label api-review

caesarxuchao commented:
/remove-sig api-machinery

msau42 commented on pkg/apis/storage/validation/validation_test.go:
```diff
@@ -1666,6 +1666,8 @@ func TestCSIDriverValidation(t *testing.T) {
 	podInfoOnMount := true
 	notPodInfoOnMount := false
 	notRequiresRepublish := false
+	storageCapacity := true
```
Should the test also toggle the feature flag on/off?

msau42 commented on pkg/registry/storage/rest/storage_storage.go:
```diff
@@ -119,6 +119,15 @@ func (p RESTStorageProvider) v1beta1Storage(apiResourceConfigSource serverstorag
 	storage["csidrivers"] = csiDriverStorage.CSIDriver
 
+	// register csistoragecapacity if CSIStorageCapacity feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIStorageCapacity) {
```
According to the comment [here](https://github.com/kubernetes/kubernetes/pull/99641/commits/43e1dd87e167233f096d39dd1249b492506b8dd8#r587487569) it seems like we shouldn't feature gate this?

liggitt replied:
We intentionally did not add support for individual resource enablement/disablement to `--runtime-config`, but kept enablement scoped to the group/version level. Verify-scripts and CI processes in kube expect `--runtime-config=api/all=true` to enable all APIs, and this subverts that.

pohly replied:
So I simply remove the feature gate check here?

pohly replied:
For the sake of expedience I have done that in a separate commit. That way the change has an explicit commit comment explaining it and if we want the check after all, we can bring it back.

msau42 commented on staging/src/k8s.io/api/storage/v1alpha1/types.go:
```diff
@@ -25,6 +25,9 @@ import (
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.9
+// +k8s:prerelease-lifecycle-gen:deprecated=1.10
```
I guess we never officially deprecated this. Should it be 1.21? Can you also open up a tracking issue so we can clean this up?

thockin commented on pkg/apis/storage/types.go:
```diff
@@ -341,7 +341,7 @@ type CSIDriverSpec struct {
 	// capacity information has been published.
 	//
-	// This is an alpha field and only available when the CSIStorageCapacity
+	// This is a beta field and only available when the CSIStorageCapacity
```
We have a new `+featureGate=CSIStorageCapacity` tag now!

pohly replied:
Does that replace this entire "This is a beta field and ..." paragraph or do I add the tag to that?

pohly replied:
The only other place that uses it does so without `+`:
https://github.com/kubernetes/kubernetes/commit/72da0b1bb06607f3f3e067f1bb5ce329ec861e1b#r47879733

pohly replied:
I've left this out because it didn't seem to make any difference in the generated files.

thockin replied:
It's new, we don't have a lot of examples yet, and we are not consuming it yet, but we are encouraging people to start setting it.

pohly replied:
There seems to be some confusion about the tag. You are saying it should be `+featureGate=CSIStorageCapacity` whereas @nikhita said in https://github.com/kubernetes/kubernetes/pull/98277#discussion_r582558737 that it should be `featureGate=CSIStorageCapacity`.

pohly replied:
Using plain text (as in the current PR) and then adding the tag with `+featureGate` make sense to me, though, so I don't mind adding that.

msau42 replied:
@nikhita can you clarify the format we should use here?

pohly replied:
FWIW, the reason why I prefer `+featureGate` is that it does not end up in the human-readable text like the `featureGate=LoadBalancerClass` in https://github.com/kubernetes/kubernetes/blob/55a5d4faf38b8cc9f65c301b9e16278048336bac/staging/src/k8s.io/api/core/v1/types_swagger_doc_generated.go#L2260

That's what we have the full sentence for.

pohly replied:
I think we can trust Tim's guidance here. He's formalizing it in a KEP:
https://github.com/kubernetes/enhancements/issues/2558

pohly replied:
@nikhita said in https://kubernetes.slack.com/archives/C5P3FE08M/p1615188110025800?thread_ts=1615058923.018400&cid=C5P3FE08M:

> I think that should be fine because the comment lines with +featureGate will be ignored by the generators, etc now. We can update it after the KEP is finalized.

thockin replied:
My intention was for it to be metadata:  `+featureGate`.  If we ultimately decide that was wrong, we can always change it, but I suspect this is better

msau42 replied:
It was added to the v1beta types.go. @thockin do you want to add it to the internal types too?

There was also question about the semantic of this tag. Does it refer to the feature gate controlling:
- whether or not values are persisted in the API server (ie we drop values if the feature is disabled)
- whether or not this api is functional (the object gets created, but nothing happens because the feature gate is turned off in the consuming components)

pohly replied:
It is meant to be processed when generating code for OpenAPI documentation, so I'd expect that it is only relevant in the external API.

Regarding the semantic: this tag is still WIP. I expect that sorting out such details will need further discussion during review of https://github.com/kubernetes/enhancements/pull/2560

thockin commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -299,7 +299,7 @@ func New(plArgs runtime.Object, fh framework.Handle) (framework.Plugin, error) {
 		capacityCheck = &scheduling.CapacityCheck{
 			CSIDriverInformer:          fh.SharedInformerFactory().Storage().V1().CSIDrivers(),
-			CSIStorageCapacityInformer: fh.SharedInformerFactory().Storage().V1alpha1().CSIStorageCapacities(),
+			CSIStorageCapacityInformer: fh.SharedInformerFactory().Storage().V1beta1().CSIStorageCapacities(),
```
What happens if the scheduler is updated before the apiserver (or before 1 of the apiservers) and ends ups askig for beta which doesn't exist?  Is that a problem or will it recover?

Or do you need to leave this on v1alpha1 until next release, at which point you can switch?

pohly replied:
> What happens if the scheduler is updated before the apiserver (or before 1 of the apiservers) and ends ups askig for beta which doesn't exist? Is that a problem or will it recover?

This is indeed a bit problematic, depending on what is expected. kube-scheduler waits for all caches to sync (cmd/kube-scheduler/app/server.go), which fails for the beta API if not supported by the kube-apiserver. kube-scheduler then waits and does not schedule anything. It will recover once the apiserver also gets updated. Is that acceptable?

Just to be sure that this really recovers, I tried it:
- deploy cluster with v1alpha1 API
- update kube-scheduler with v1beta1 API
- create pod with unbound volume, late binding and storage capacity tracking -> stuck, as expected
- update apiserver with v1beta1 API
- kube-scheduler gets the storage capacity information, pod starts

> Or do you need to leave this on v1alpha1 until next release, at which point you can switch?

Using v1alpha1 wouldn't help: in a normal cluster, that API group won't be available in kube-apiserver by default, so updating kube-scheduler before kube-apiserver will still stop scheduling.

But is this really a problem? https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/#manual-deployments explicitly instructs to upgrade etcd and kube-apiserver first. If admins don't follow that advice, some breakage has to be expected and isn't too bad in this case (will recover, log output about watch explains what's failing).

pohly replied:
> Or do you need to leave this on v1alpha1 until next release, at which point you can switch?

My other motivation for using the v1beta1 API was that I intended to add MaximumVolumeSize there, and only there. But that actually then caused test failures during roundtripping and munnerz convinced me that the new field should be added also to v1alpha1 (https://kubernetes.slack.com/archives/C0EG7JC6T/p1614956222111400?thread_ts=1614955402.110000&cid=C0EG7JC6T).

Therefore v1alpha1 and v1beta1 are now functionally equivalent and we could use it in kube-scheduler without loss of functionality. In case of version skew against an older kube-apiserver it simply wouldn't get the MaximumVolumeSize, which is fine.

But my point about the alpha API group not being enabled still stands. Only clusters where the CSIStorageCapacity API was already enabled prior to 1.21 will be able to use kube-scheduler 1.21. Is that enough reason to change the PR or should I keep it as-is, i.e. use v1beta1 in kube-scheduler?

msau42 replied:
Since apiserver needs to be upgraded first, I think it is fine to switch the scheduler to use v1beta1 (and the fact that v1alpha1 needs explicit opt-in)

pohly committed:
CSIStorageCapacity: add MaximumVolumeSize to v1beta1 API

pohly committed:
CSIStorageCapacity: check MaximumVolumeSize during scheduling

If available, then the MaximumVolumeSize is a better indicator whether
creating a volume has a chance to succeed than the total (?) Capacity,
which is potentially larger and less well-defined.

pohly commented:
The MaximumVolumeSize made it into the CSI spec 1.4.0-rc1 and thus almost certainly will be in 1.4.0. Therefore it makes sense to add the corresponding field also to the Kubernetes API and to start using it, if populated by a CSI driver. I added two commits for that, one which adds the field and one which uses it.

Note that we do not actually need to import the new CSI spec bindings. I was a bit confused about that earlier and thought that we would have to, but that's only true for external-provisioner (producing CSIStorageCapacity objects with the field) and csi-driver-host-path (simulating maximum volume size).

I contemplated adding the MinimumVolumeSize. But then we would either have a field which is completely unused or start using `ResourceRequirements.Limits`, something that (as far as I know) has been ignored so far. In the end I decided against adding that field.

pohly commented:
/retest

pohly committed:
CSIStorageCapacity: register types unconditionally

That the object was registered depending on the feature gate was
called out as unusual during the 1.21 review. Previously, all beta
storage APIs were unders such feature gate checks, but its better to
drop that to be consistent with the rest of Kubernetes.

thockin reviewed (commented):
From an API POV this looks ok, though I admit I am less familiar with the lifecycle-gen tag.

I'll approve, but @msau42 holds the LGTM

/approve

pohly commented:
> I'll approve, but @msau42 holds the LGTM

Thanks! I'll add `+featureGate` for that so that it is in place when she gets to look at the PR.

pohly committed:
CSIStorageCapacity: add +featureGate=CSIStorageCapacity

The tag is currently mostly ignored by the tooling, but its usage is
encouraged (https://github.com/kubernetes/kubernetes/pull/99641#discussion_r588078444).

pohly commented:
/retest

msau42 commented on test/e2e/storage/csi_mock_volume.go:
```diff
@@ -31,7 +31,7 @@ import (
 	v1 "k8s.io/api/core/v1"
 	storagev1 "k8s.io/api/storage/v1"
-	storagev1alpha1 "k8s.io/api/storage/v1alpha1"
+	storagev1beta1 "k8s.io/api/storage/v1beta1"
```
The feature tag should be removed so it always runs

pohly replied:
Funny, this also occured to me this morning during breakfast :sweat_smile: 

Done.

pohly committed:
CSIStorageCapacity: enable E2E tests by default

With the feature now in beta, tests should run in the normal jobs.

pohly commented:
/retest

pohly commented:
It looks like the 1.21 milestone still needs to be set for this PR, too.

msau42 commented on staging/src/k8s.io/api/storage/v1beta1/types.go:
```diff
@@ -569,3 +571,98 @@ type CSINodeList struct {
+// - such an object exists, but the capacity is zero
+//
+// The producer of these objects can decide which approach is more suitable.
+type CSIStorageCapacity struct {
```
Can we document feature gate requirements here too?

pohly replied:
This object kind is available regardless of the feature gate. This is the key difference to other, earlier alpha and beta object kinds which were only registered when their flag was set.

What we can say here is "This information is consumed by the kube-scheduler if the CSIStorageCapacity beta feature gate is enabled and a CSI driver opts into capacity-aware scheduling with CSIDriver.StorageCapacity.".

Is that okay?

pohly replied:
As discussed on chat, I added that sentence.

msau42 reviewed (commented):
Just one more comment, otherwise lgtm

msau42 commented:
/milestone v1.21

pohly commented:
> pull-kubernetes-verify failed.

I need to rebase and regenerate one file... working on it.

msau42 commented:
/lgtm

The feature-gate tagging discussion can continue offline and we can update the tags once the KEP is merged.

msau42 commented:
Oh @pohly can you also update the release note saying the alpha capacity objects are also deprecated?

pohly commented:
> Oh @pohly can you also update the release note saying the alpha capacity objects are also deprecated?

Done.

pohly commented:
/retest

fedebongio commented:
/triage accepted
